uplink/eestream/decode.go

```diff
@@ -192,6 +192,8 @@ func (dr *decodedRanger) Size() int64 {
 
 func (dr *decodedRanger) Range(ctx context.Context, offset, length int64) (_ io.ReadCloser, err error) {
 	defer mon.Task()(&ctx)(&err)
+
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
As far as I understand, with this change, closing the DecodeReader will cancel the context for all piece readers. This happens right before calling the Close() method of the piece readers.

Will it still be possible for `Download.Close()` to read the read via `client.stream.Recv()` if the context is already canceled?

Reply:
Ah, right.. probably not.

Reply:
`client.stream.Recv()` will return a context canceled error in some cases, but we are ignoring these errors in `Download.Close()`

Reply:
The question is - will it return the actual error that happened instead of "context canceled"?

Reply:
No, I added a function `ignoredCanceled(err)` that returns nil if `errs2.IsCanceled(err)` is true. Which means if there is some other error, it will be ignored. I do not know what circumstances could cause a context canceled error + some other error wrapped together.

Reply:
Note from meeting - create ticket for making sure we can properly read errors in download.Close() in the case of a canceled context + some other error we do not want the context cancelation to obscure.

Reply:
I created https://github.com/storj/storj/pull/2771 which should handle the errors better.